### ISSUE TITLE
Add support for inline scripts and styles using the nonce

### ISSUE THREAD

**Issue by RehanSaeed** [NWebsec/NWebsec#54]
The <a href="http://www.w3.org/TR/CSP2">W3C</a> page for CSP 2.0 talks about using a nonce to enable in-line styles and scripts. In my personal opinion, this is the biggest hurdle for adoption of CSP by the mainstream. Modernizr <a href="https://github.com/Modernizr/Modernizr/pull/1263">STILL</a> does not support CSP because of this.

It would be great to add support for this feature to NWebSec. An MVC HTML helper could be created that worked a little like the Anti-forgery token does now.

My only reservation with this feature is that existing browsers don't support CSP 2.0 but do support CSP 1.0. They will block the in-line styles and scripts making this feature possibly useless there is some workaround or those browsers reach the end of their lives.

**Comment by klings** [comment]
NWebsec supports it! You'll find it in the docs explaining how to [configure Csp](https://github.com/NWebsec/NWebsec/wiki/Configuring-Content-Security-Policy#script-and-style-nonces-through-htmlhelpers).

There's no support for script hashes yet though. Unfortunately, it would require a bit of work. If someone has a legitimate use for it and files an issue on it, I might add it.

A comment on the CSP nonce, the intention with the nonce is to make it possible to add CSP to existing apps that already contain inline scripts - without requiring a rewrite of the app. The idea is to combine the nonce with 'unsafe-inline'. CSP 1.0 UAs will execute the inline script as before (insecure but backwards compatible), but CSP 2.0 UAs will disregard 'unsafe-inline' when they see the nonce, and only execute inline scripts with the nonce set. This gives an "upgrade" path for existing apps, and they can benefit from CSP 2 without requiring a massive rewrite to get rid of inline JS.

Note that nonces do not work well with output caching. That's one of the use cases for hashing, but there are some tricky problems related to ASP.NET caching that needs to be solved to make it work in a robust way.

**Comment by RehanSaeed** [comment]
Very good read. Thanks! I had not thought of caching, I can see why using the nonce is not a great method of using in-line scripts. Hashes might be the way to go but is more complicated to setup. I would definitely put my vote in for their inclusion into NWebSec, it could be one way to get third party libraries to work with CSP.

I've raised issues with <a href="https://github.com/Modernizr/Modernizr/pull/1263">Modernizr</a> and <a href="http://webessentials.uservoice.com/forums/140520-general/suggestions/6665824-browser-link-support-for-content-security-policy">Visual Studio Browser Link</a> which both use in-line scripts and fail CSP. So far, the issues have been ignored. Hashes might be a way around that.

The CSP 2.0 upgrade path is interesting, I assume the same would occur for hashes as well as nonces.

I've upgraded <a href="https://github.com/RehanSaeed/ASP.NET-MVC-Boilerplate">ASP.NET MVC Boilerplate</a> with the changes in CSP 2.0 and learned a lot in the process. The next thing on my list is a couple of NWebSec blog posts...finally.
